### PR TITLE
Feature : Collateralization oracle

### PR DESCRIPTION
**Pull request by eswak**
Hi,

Here is a pull request to add a Collateralization Oracle as discussed [on the governance forums](https://tribe.fei.money/t/ethreweightqueuer-proposal-to-improve-reweight-mechanism-maintain-peg/1521/41).

This contract is pretty straightforward & mainly composed of pure functions : 
- It interacts with `EthUniswapPCVDeposit` to get the `totalValue()` and `getReserves()`.
- Circulating FEI supply is defined as FEI `totalSupply()` minus FEI in the Uniswap pool owned by `EthUniswapPCVDeposit`.
- `UniswapOracle` is used for the ETH/USD current price.
- `UniswapOracle` is used for the ETH/FEI current price (234b68f16b803c0dba7785565173dc8ca40c8159).
- It should be upgraded as PCV is deployed elsewhere in the future (as in #95).

I also implemented a small test suite inspired from the one of `UniswapOracle` :
Can be run with ` npm run test -- -g CollateralizationOracle` 
image

Also tested on a local mainnet fork, with results consistent with what's observed on https://app.fei.money/analytics (see comments below)

Best,

### CODE REVIEW

File: contracts/oracle/CollateralizationOracle.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        ethPriceOracle = UniswapOracle(_ethPriceOracle);
+        ethUniswapPCVDeposit = EthUniswapPCVDeposit(_ethUniswapPCVDeposit);
+    }
+
+    /// @notice get the current circulating suply of FEI
+    /// @return number of FEI in circulation
+    function _circulatingFei() internal view returns(uint256) {
+      (uint256 feiInPool, uint256 ethInPool) = ethUniswapPCVDeposit.getReserves();
+      uint256 ethPcv = ethUniswapPCVDeposit.totalValue();
+      uint256 feiPcv = Decimal.ratio(ethPcv, ethInPool).mul(feiInPool).asUint256();
```
Comment (Joeysantoro) [inline contracts/oracle/CollateralizationOracle.sol:39]: Using the spot reserves of Uniswap would be vulnerable to flash manipulation. I'd suggest using an ETH-FEI TWAP oracle and then back out manipulation resistant numbers for the reserves by doing `k = feiInPool * ethInPool`
`resistantEthInPool = sqrt(k/ethFeiOraclePrice)` and `resistantFeiInPool = k/resistantEthInPool`

Here is the address for our current ETH-FEI TWAP: 0xB58cbCb1e98c7F3771847cc2D730bF9589b534e4. Its 30 min so would only be more useful for less price sensitive applications

**Review by Joeysantoro — CHANGES_REQUESTED** [review]
This is awesome thanks for coding it up so fast!

One final note is it is good to reduce the number of external contract calls as much as possible to save on gas. One call to read should only hit every external call once in the entire flow.

File: contracts/oracle/CollateralizationOracle.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "./UniswapOracle.sol";
+import "../refs/CoreRef.sol";
+import "../external/SafeMathCopy.sol";
+import "../pcv/EthUniswapPCVDeposit.sol";
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:7]: I used `UniswapPCVDeposit` that is also a `IUniRef` to be able to call `getReserves()`.

File: contracts/oracle/CollateralizationOracle.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "./UniswapOracle.sol";
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:4]: done

File: contracts/oracle/CollateralizationOracle.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /// @notice get the current circulating suply of FEI
+    /// @return number of FEI in circulation
+    function _circulatingFei() internal view returns(uint256) {
+      (uint256 feiInPool, uint256 ethInPool) = ethUniswapPCVDeposit.getReserves();
+      uint256 ethPcv = ethUniswapPCVDeposit.totalValue();
+      uint256 feiPcv = Decimal.ratio(ethPcv, ethInPool).mul(feiInPool).asUint256();
+
+      return Decimal.from(fei().totalSupply()).sub(feiPcv).asUint256();
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:41]: done

File: contracts/oracle/CollateralizationOracle.sol (hunk 1 above)
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:44]: done

**Comment by eswak** [comment]
I changed the implementation of `_collateralizationRatio()` to make the external calls only once.

But since we introduced the `ethFeiPriceOracle`, there are still 5 external calls to get the collateral ratio. Should we have an internal variable to keep the last value, and rely more on update() calls ?

On a related topic : is the code of the webapp (https://app.fei.money/analytics) available somewhere ? The analytics page could be updated to read the oracle.

**Comment by Joeysantoro** [comment]
> On a related topic : is the code of the webapp (https://app.fei.money/analytics) available somewhere ? The analytics page could be updated to read the oracle.

The webapp code is not available but we could update it! Have you tested this oracle on a mainnet fork?

**Comment by eswak** [comment]
Note : to be updated after #95 is merged (`EthReserveStabilizer` and `EthPCVDripper` will have non-zero balance of ETH that must be counted as part of the PCV).

File: contracts/oracle/CollateralizationOracle.sol [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    function collateralizationRatio() external view returns(uint256) {
+      (Decimal.D256 memory ratio,) = _collateralizationRatio();
+      return ratio.mul(1000000000000000000).asUint256();
+    }
+
+    /// @notice get a boolean wheter  current collateralization ratio
+    /// @return true if the protocol is currently overcollateralized
+    function overCollateralized() external view returns(bool) {
+      (Decimal.D256 memory ratio,) = _collateralizationRatio();
+      return ratio.mul(1000000000000000000).asUint256() > 1000000000000000000;
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:153]: done

File: contracts/oracle/CollateralizationOracle.sol [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    /// @return price in USD per ETH
+    function ethUsd() external view returns(uint256) {
+      return _ethUsd();
+    }
+
+    /// @notice get the current collateralization ratio
+    /// @return collateralization ratio
+    /// @dev price is to be denominated in percentage, with 18 decimals (1.25e18 for 125%)
+    function collateralizationRatio() external view returns(uint256) {
+      (Decimal.D256 memory ratio,) = _collateralizationRatio();
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:145]: done

File: contracts/oracle/CollateralizationOracle.sol [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    /// @notice updates the oracle price
+    /// @return true if oracle is updated and false if unchanged
+    function update() external override whenNotPaused returns (bool) {
+        bool updated = ethUsdPriceOracle.update();
+        updated = updated || ethFeiPriceOracle.update();
+
+        (Decimal.D256 memory ratio,) = _collateralizationRatio();
+
+        emit Update(ratio.asUint256());
```
[the remark this answers is not in the document]
Comment (eswak) [reply contracts/oracle/CollateralizationOracle.sol:112]: done (the `UniswapOracle` always return true, that's why I didn't bother, but their implementation may change)

**Pull request closed without merging**